drivers/rkt/driver.go

```diff
@@ -0,0 +1,991 @@
+// +build linux
+
+package rkt
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"syscall"
+	"time"
+
+	appcschema "github.com/appc/spec/schema"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/client/driver/env"
+	dstructs "github.com/hashicorp/nomad/client/driver/structs"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/drivers/shared/eventer"
+	"github.com/hashicorp/nomad/plugins/base"
+	"github.com/hashicorp/nomad/plugins/drivers"
+	"github.com/hashicorp/nomad/plugins/drivers/utils"
+	"github.com/hashicorp/nomad/plugins/shared/hclspec"
+	rktv1 "github.com/rkt/rkt/api/v1"
+	"golang.org/x/net/context"
+
+	"encoding/json"
+	"io/ioutil"
+	"math/rand"
+	"strconv"
+
+	"github.com/hashicorp/consul-template/signals"
+	"github.com/hashicorp/nomad/client/driver/executor"
+)
+
+const (
+	// pluginName is the name of the plugin
+	pluginName = "rkt"
+
+	// fingerprintPeriod is the interval at which the driver will send fingerprint responses
+	fingerprintPeriod = 30 * time.Second
+
+	// minRktVersion is the earliest supported version of rkt. rkt added support
+	// for CPU and memory isolators in 0.14.0. We cannot support an earlier
+	// version to maintain an uniform interface across all drivers
+	minRktVersion = "1.27.0"
+
+	// rktCmd is the command rkt is installed as.
+	rktCmd = "rkt"
+
+	// rktNetworkDeadline is how long to wait for container network to start
+	rktNetworkDeadline = 1 * time.Minute
+)
+
+var (
+	// pluginInfo is the response returned for the PluginInfo RPC
+	pluginInfo = &base.PluginInfoResponse{
+		Type:             base.PluginTypeDriver,
+		PluginApiVersion: "0.0.1",
+		PluginVersion:    "0.1.0",
+		Name:             pluginName,
+	}
+
+	// configSpec is the hcl specification returned by the ConfigSchema RPC
+	configSpec = hclspec.NewObject(map[string]*hclspec.Spec{
+		"volumes_enabled": hclspec.NewDefault(
+			hclspec.NewAttr("volumes_enabled", "bool", false),
+			hclspec.NewLiteral("true"),
+		),
+	})
+
+	// taskConfigSpec is the hcl specification for the driver config section of
+	// a taskConfig within a job. It is returned in the TaskConfigSchema RPC
+	taskConfigSpec = hclspec.NewObject(map[string]*hclspec.Spec{
+		"image":              hclspec.NewAttr("image", "string", true),
+		"command":            hclspec.NewAttr("command", "string", false),
+		"args":               hclspec.NewAttr("args", "list(string)", false),
+		"trust_prefix":       hclspec.NewAttr("trust_prefix", "string", false),
+		"dns_servers":        hclspec.NewAttr("dns_servers", "list(string)", false),
+		"dns_search_domains": hclspec.NewAttr("dns_search_domains", "list(string)", false),
+		"net":                hclspec.NewAttr("net", "list(string)", false),
+		"port_map":           hclspec.NewAttr("port_map", "map(string)", false),
```

Comment:
Should become: `"port_map":           hclspec.NewBlockAttrs("port_map", "string", false),` once this gets merged and r-clientv2 rebases: https://github.com/hashicorp/nomad/pull/4790